### ISSUE TITLE
Multiprocess writes using `to_hdf5`

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#3074]
After seeing @stuartarchibald's [post on gitter]( https://gitter.im/dask/dask?at=5a5cd598290a1f45618f91b3 ), I was a bit curious how `to_hdf5` actually worked and whether it would be viable to call from multiple processes.

Turns out [`to_hdf5` just calls `store` underneath the hood]( https://github.com/dask/dask/blob/0.16.1/dask/array/core.py#L3390 ) while [holding the file open locally]( https://github.com/dask/dask/blob/0.16.1/dask/array/core.py#L3385 ). The biggest issue of course is that the file is held open in write mode on the scheduler (meaning nothing else can write to the HDF5 file without corrupting it). Since the type of locking was not specified and `store` uses a [`threading.Lock` by default]( https://github.com/dask/dask/blob/0.16.1/dask/array/core.py#L837-L840 ), this is also incompatible with the multiprocessing use case (i.e. `threading.Lock` cannot be serialized/locked across processes). IOW the current implementation of `to_hdf5` is not friendly even for locked parallel writes.

That said, it should be feasible to change `to_hdf5`'s behavior to be more friendly for storage from multiple processes (though it will still need to be locked). In this case, it would still create the datasets initially (as it already does), but then would close the file. Instead of passing raw HDF5 Datasets as targets, a wrapper class would need to be used (to allow for pickling). The wrapper class would need to provide a `__setitem__` method that would open the HDF5 file and write to the HDF5 Dataset at the selection specified in a process safe manner (probably with `locket.lock_file`). Ideally it would provide a `__getitem__` method as well. Doing this should allow for HDF5 file to be written to in parallel. This assumes the filesystem is very robust and syncing changes between different nodes.

An alternative to this proposal that would avoid locking would be to have data serialized back to the scheduler, which then writes each piece to the HDF5 file as it arrives. This would avoid the overhead of the previous strategy (and any potential issues locking) by guaranteeing only one process ever opens the HDF5 file and keeps open until everything is written. This strategy would continue to work well for non-parallel use cases with arguably less overhead than is present now. The only thing to do before writing out all results would be to optimize the graphs since `store` would no longer be used.

**Comment by rzu512** [comment]
Can I avoid this problem if I use the MPI scheduler?
h5py [says](http://docs.h5py.org/en/latest/mpi.html) it supports parallel hdf5.

**Comment by jakirkham** [comment]
I don't really have much experience with Parallel HDF5. However, from what I have read, it should be possible to use MPI to write in parallel. It would probably be too difficult to enforce collectivity, but that doesn't seem to be required for a basic write. That said, we would still need to fix the lock issue.

We probably cannot generally assume that most people wanting to use `to_hdf5` are using MPI (there are a few hoops for the average user). That said, I agree there would be a lot of value in making sure that the solution to this problem remains MPI friendly. We could add a flag to disable locking (with locking enabled by default). Alternatively we can try to detect whether the user is running with MPI. I have a slight preference for the former (flag-based) solution, but could be convinced otherwise.

Any other suggestions about improving the MPI friendliness of `to_hdf5`?

**Comment by mrocklin** [comment]
> That said, it should be feasible to change to_hdf5's behavior to be more friendly for storage from multiple processes (though it will still need to be locked). In this case, it would still create the datasets initially (as it already does), but then would close the file. Instead of passing raw HDF5 Datasets as targets, a wrapper class would need to be used (to allow for pickling). The wrapper class would need to provide a __setitem__ method that would open the HDF5 file and write to the HDF5 Dataset at the selection specified in a process safe manner (probably with locket.lock_file). Ideally it would provide a __getitem__ method as well. Doing this should allow for HDF5 file to be written to in parallel. This assumes the filesystem is very robust and syncing changes between different nodes.

This is a decent option.  We could also pass the path and datapath as part of the task.  We might also want to cache the h5py file objects in some sort of LRU cache.

It's worth noting that XArray has solved this problem.  Honestly the fact that XArray exists is probably why no one has spent much time on this topic historically.  We should still have a pure-dask.array solution though.  We might want to look at what they do.

> An alternative to this proposal that would avoid locking would be to have data serialized back to the scheduler, which then writes each piece to the HDF5 file as it arrives. 

I'm -1 on this.  It doesn't scale well and the scheduler shouldn't get bogged down.

> Can I avoid this problem if I use the MPI scheduler?  h5py says it supports parallel hdf5.

No.  There is no MPI scheduler.  As stated in [the docs](http://dask.pydata.org/en/latest/setup/hpc.html#using-mpi) dask-mpi only uses MPI to set up dask workers, and not for inter-worker communication.

**Comment by mrocklin** [comment]
Generally it would be great to improve `to_hdf5`.  There is some good discussion on this topic on this XArray issue.  I'm pointing specifically to my comment there where I propose a solution:  https://github.com/pydata/xarray/issues/798#issuecomment-199545836

**Comment by jakirkham** [comment]
Thanks for the thoughts.

An LRU cache definitely makes sense for reading. I think that is what the `h5pickle` project did. Believe the author was interested in contributing that back to Dask. So we definitely can follow-up on that.

Am a little nervous using an LRU cache for writing though. Have corrupted an HDF5 file before during testing by merely opening the file in append mode and reading from it in multiple processes. So would certainly appreciate any elucidation you might have on this point (assuming I'm understanding the implication correctly).

Would be remiss if I didn't mention @martindurant's comment about `dask.dataframe` writing from multiple workers behaving correctly. Perhaps we don't need `lockfile` and can just [copy that strategy]( https://github.com/dask/dask/blob/0.16.1/dask/dataframe/io/hdf.py#L163-L180 ) over as he suggested.

Agree that writing from the scheduler is suboptimal. However it can be handy if the underlying file system is a bit unreliable. Though I agree it is not the best direction.

**Comment by mrocklin** [comment]
If we wanted to write from a single process sequentially then I would
volunteer one of the workers rather than the scheduler.

I wonder if we can guarantee good behavior if we chunk the dask.array along
the hdf5 dataset chunks.  Perhaps that avoids the need for an inter-process
lock and we can just get away with intra-process locks?

**Comment by jakirkham** [comment]
Good point. Agreed. Am now wondering if we should persist the data and then have one worker write things out as soon as they complete.

Don't think I'm following this second proposal. How are we handling the file handle constraints?

**Comment by mrocklin** [comment]
My hope is that if we only write cleanly across blocks (no two processes
will write to the same block) then maybe we can avoid locking between
processes and allow multi-process writes.  I don't know though.

**Comment by stuartarchibald** [comment]
I've just caught up with this, thanks @jakirkham for opening the ticket. Indeed, `to_hdf5()` just calls `store()` and does not work in the distributed context due to the use of threadlocals for the lock, I tried a `SerializableLock` and then IIRC the hdf5 object won't serialize instead. I guess I anticipated that there would be some coordinated write taking place whereby either a lock was passed around between processes and then file opening/closing plus the actual write happening from those processes. In practice I resorted to just materialising the `dask.array` data chunk-wise on a single process and writing those chunks to file as they appeared (I'd love a chunks API, but that's another story!).

The following is based on in part memory and in part a bit of looking through docs...

The hdf5 FAQ (https://support.hdfgroup.org/HDF5/hdf5-quest.html) mentions a few useful things:
1. Concurrent access from multiple processes is fine if reading, if writing SWMR (single-write-multiple-read, pronounced "swimmer" apparently) can be used, but note the "single-write". https://support.hdfgroup.org/HDF5/hdf5-quest.html#gconc1 (http://docs.h5py.org/en/latest/swmr.html, www.hdfgroup.org/HDF5/docNewFeatures/NewFeaturesSwmrDocs.html).
2. That MPI with MPI-I/O is needed for Parallel HDF5. https://support.hdfgroup.org/HDF5/hdf5-quest.html#pneed

So in terms of parallel writes to hdf5, I don't think this is possible without the use of `MPI` as it is `MPI-I/O` that coordinates this ability. Further, concurrent writes from multiple processes are not supported, but single writer multiple reader (SWMR) is. I think therefore locked and serialised write coordination would be needed across processes and but reads could be done simultaneously.

With a bit more thought, in terms of write from multiple processes, I guess one (somewhat complicated) way to achieve a faked-up MPI-I/O like parallel write would be for each process to open a hdf5 file for its chunk of data and write to it lock free. Then use that the hdf5 spec supports references to external data or Virtual Datasets (VDS) to assemble an agglomerating hdf5 file on the scheduler thread that references the per-process written data, this acting as a transparent entry point for future reads. There are however some restrictions on the types of data that can be used for VDS/external, some more info is here https://support.hdfgroup.org/HDF5/Tutor/vds.html and here https://support.hdfgroup.org/HDF5/Tutor/property.html#storage

As to the noted corruption from read whilst in append mode, the SWMR  interface may help http://docs.h5py.org/en/latest/swmr.html.

**Comment by mrocklin** [comment]
Note that dask.distributed also has a distributed lock:
http://distributed.readthedocs.io/en/latest/api.html#distributed.Lock

**Comment by stuartarchibald** [comment]
Thanks @mrocklin for https://github.com/dask/dask/issues/3074#issuecomment-359031283, perhaps I ought to try `store()` with that too. I'd still really like lock free parallel writes though!

**Comment by jakirkham** [comment]
> I tried a `SerializableLock` and then IIRC the hdf5 object won't serialize instead.

Right. So my thinking here is that we wrap up the opening and closing of the HDF5 File such that it only occurs when storing happens. This way we are only pickling the filename, datasetname, and some slice info about where to store each chunk.

> In practice I resorted to just materialising the `dask.array` data chunk-wise on a single process and writing those chunks to file as they appeared

That was something Matt and I were talking about earlier in this thread. We could also submit this as a job, which would free you up to do other things while the result is serialized.

> (I'd love a chunks API, but that's another story!).

There are a few tricks of the trade for this sort of stuff. Maybe we could do with some internal developer docs about how to work with chunks.

> ...one (somewhat complicated) way to achieve a faked-up MPI-I/O like parallel write would be for each process to open a hdf5 file for its chunk of data and write to it lock free.

Had also given this some thought more recently. We might be able to employ some tricks copying these chunks back into one HDF5 file also. Maybe a little less tricky than VDS.

> I'd still really like lock free parallel writes though!

Personally have not found HDF5 satisfying in this regard. Would make another mention for Zarr here as that seems like a better fit. It was designed with Dask usage in mind. So avoids these pickling and locking issues that HDF5 bumps into.

**Comment by stuartarchibald** [comment]
I've found `MPI-I/O` fine for parallel read/write but this is indeed an entirely different system. I appreciate that `Zarr` is almost certainly a better fit, but sometimes `HDF5` is a constraint applied by external factors. I'm of the view that if something can be done to make `to_hdf5` work better with distributed in the short term, even if that's just making it so that the writes work out of the box albeit in practice they are serial, then that is a good intermediate step. Long term I'd be interested in hearing about ideas of copying chunks into a single hdf5 file and how that compares to VDS.

**Comment by jakirkham** [comment]
> I've found `MPI-I/O` fine for parallel read/write but this is indeed an entirely different system.

I'm sure my experience would be different if I were using MPI-I/O.

> I appreciate that `Zarr` is almost certainly a better fit, but sometimes `HDF5` is a constraint applied by external factors.

I can definitely appreciate that. We still use HDF5 as a storage format. It remains a nice format that lots of people in different languages can read and write easily and there are some standards in my research community that have been built around it. So this is unlikely to change for me either.

However when it comes to computing things with Dask, we convert everything into Zarr and only use Zarr until we are done. At which point it goes back into some HDF5 file that we share and/or store longer term. IOW Zarr is internal to our computational pipelines with the option to export to HDF5 wherever needed (typically the end).

There has been a lot of work in Zarr `master` (soon to be 2.2.0) to make sure that conversion between Zarr and HDF5 is pretty painless. Not really knowing your constraints, I don't know if this strategy would be a good fit for your use case. That said, if you think it might, would certainly encourage you to explore it further and ask questions as they arise.

> I'm of the view that if something can be done to make `to_hdf5` work better with distributed in the short term, even if that's just making it so that the writes work out of the box albeit in practice they are serial, then that is a good intermediate step.

Generally I agree. Unfortunately, for me, short term would probably be in a month. Though I would be supportive of you or anyone pushing ahead on one of the ideas we have discussed.

> Long term I'd be interested in hearing about ideas of copying chunks into a single hdf5 file and how that compares to VDS.

Sure. I guess let's wait and see how the short term solution goes.

**Comment by rzu512** [comment]
Can this approach handle a Dask array that has `np.nan` in chunks?
The method mentioned [here](https://clouds.eos.ubc.ca/~phil/courses/parallel_python/03_dask_and_zarr.html) seems to not work for this case.

**Comment by jakirkham** [comment]
Ultimately working with Arrays of unknown shape would be difficult. How would you propose allocating an adequately sized HDF5 Dataset?

**Comment by jakirkham** [comment]
We could also consider an entirely different approach if we could guarantee there is a `Client`. Might provide some impetus to think about issue ( https://github.com/dask/dask/issues/3237 ). ;) Namely we could `persist` the Dask Array and then pass all the Futures to one Dask job, which could then write them out as they arrive. This would avoid having a file lock and would allow the HDF5 file to only be opened once by one worker. Other safety measures could be added here like writing to a temp file that gets moved to the right location only when it is done. Of course this wouldn't allow parallel writes; though, neither does the strategy in PR ( https://github.com/dask/dask/pull/3179 ).

cc @shoyer @jhamman

**Comment by jakirkham** [comment]
On a different note, took a look at what support for VDS was in h5py and it seems there has been an attempt to add Python support for this, but it is incomplete. That said, we could certainly explore a simple solution along these lines (an HDF5 file per chunk) as Dask can already help do the legwork for this sort of thing without too much trouble. This would allow parallel reading and writing as a direct consequence of using multiple files. Though keeping track of these files would come with some overhead.

**Comment by jhamman** [comment]
From, my perspective, streaming computed arrays when using `distributed` to a single writer would be a nice compromise here. I'm not sure what machinery would be needed to make that happen but I can vouch for its potential application. 

In terms of virtual datasets, I'm headed this way with xarray as well. In our case we can use `save_mfdataset(datasets, paths)` to enable parallel writing to separate files. `datasets` would be some partitioned version of a single `xarray.Dataset` made up of dask arrays.

**Comment by jakirkham** [comment]
It’s worth noting that Zarr has support for copying data into HDF5 files.

ref: http://zarr.readthedocs.io/en/stable/api/convenience.html#zarr.convenience.copy

**Comment by rddtz** [comment]
I recently get involved in a project that needed to save dask arrays
to hdf5 files in a distributed system. As dask.to_hdf5 function
doens't worked due to pickling issues, we did a work-around using HDF5
Virtual Datasets (https://docs.h5py.org/en/latest/vds.html). Our idea
was to create one separated file for each chunk and then merging it in
one single file using the Virtual Dataset capabilities, similar to
what zarr does (also with one file per chunk). This then enables full
parallel writing without the need for a lock. This solution also
removed the need to pickle hdf5 objects since we ship several tasks
that on its own handles the file open, writing and closing for each separated
chunk, and not send opened hdf5 objects.

We were then thinking if this method could be used to solve this issue? 
I created a PR with a simple implementation if sounds promising: https://github.com/dask/dask/pull/12309#issue-3996729623

**Comment by dcherian** [comment]
Can you provide a reproducer for the Zarr issue please?

**Comment by rddtz** [comment]
Sorry, I mispelled, it is the **`dask.to_hdf5`** function that has problem with issues.

Here is a small example of where it fails:

Start workers.
```
dask-scheduler&
dask-worker 127.0.0.1:8786 & 
dask-worker 127.0.0.1:8786 & 
```

Then this code would fail:

```
import dask.array as da
from dask.distributed import Client,wait,progress
import h5py

client = Client("127.0.0.1:8786")

chunk_shape=(5000,5000)
data_shape=(10000, 10000)
data = da.random.random(data_shape, chunks=chunk_shape)

data.to_hdf5("dask.hdf5", "data")
```

The error would be something like this with some more warnings:

```
...
    return super().dump(obj)
           ~~~~~~~~~~~~^^^^^
  File "path/to/.venv/lib/python3.13/site-packages/h5py/_hl/base.py", line 368, in __getnewargs__
    raise TypeError("h5py objects cannot be pickled")
TypeError: h5py objects cannot be pickled

```
